### PR TITLE
networkd: allow networkd to start in early boot

### LINKED ISSUE #4504: allow networkd to start in early boot
### Submission type
- [ ] Bug report
- [X] Request for enhancement (RFE)

Commit 1346b1f038 introduced an ordering of networkd to run after D-Bus. This was further refined in commit bcbca8291f. This effectively forces networkd to start in late boot (after `basic.target`), as `dbus.service` runs in late boot and networkd orders itself after that. So networkd's `DefaultDependencies=no` is effectively useless. In https://launchpad.net/bugs/1636912 the cloud-init developers asked for providing a way to make networking available in early boot, similar to what we had with Debian's ifupdown (`networking.service`).

For a similar reason (using resolved in early boot) I recently tried to make D-Bus run earlier (https://bugs.freedesktop.org/show_bug.cgi?id=98254), to get a little closer to the spirit of kdbus ("D-Bus is always available"). But discussion showed that D-Bus services are at least potentially not ready for this, so it was rejected (of course these issues would hit again if/when something like kdbus or systemd-dbus actually becomes a reality).

It seems to me that networkd is actually meant to have very little dependencies, given that you can even run it in the initrd (for iSCSI, remote root and the like). And that generally works, except for the `UseHostname: yes` case as then it needs to talk to hostnamed for setting the transient hostname.

Would it be feasible to drop the  `After=dbus.service` and instead queue/postpone the setting of the transient hostname until D-Bus starts up? The current code (commit bcbca8291f) entirely ignores the sent hostname if D-Bus isn't running, but by postponing this we would make networkd independent of dbus.service, allow it to start from initrd even, and still keep the transient hostname feature.

### PR DESCRIPTION
**Pull request by martinpitt**
If setting the received transient hostname fails because D-Bus is not (yet) up,
store the host name and corresponding Link in the Manager so that it can try
again after connecting to D-Bus.

With this the "After=dbus.service" can be dropped, which effectively forced
networkd.service to run in late boot only (dbus.service was rejected to run in
early boot in https://bugs.freedesktop.org/show_bug.cgi?id=98254).

Fixes #4504 

Note: While this makes things work when networkd starts before D-Bus, it's still not as nice as it could be. See the discussion in #4504 about replacing the 5s retry loop with a signal handler. Let me know if you prefer this, then I can work on that.

### CODE REVIEW

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -442,6 +443,12 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
                         r = link_set_hostname(link, hostname);
                         if (r < 0)
                                 log_link_error_errno(link, r, "Failed to set transient hostname to '%s': %m", hostname);
+                        else if (r == 0) {
+                                if (free_and_strdup(&link->manager->pending_hostname, hostname) < 0)
+                                    return log_oom();
+                                link->manager->pending_hostname_link = link;
```
Comment (poettering) [inline src/network/networkd-dhcp4.c]: if the link disappears after this then the pending_hostname_link will point to nowhere, and we might deref it and access invalid memory...

File: src/network/networkd-manager.c [hunk 2]
```diff
@@ -150,6 +150,16 @@ int manager_connect_bus(Manager *m) {
...
 
+        /* Do we have a pending hostname from DHCP while D-Bus wasn't up yet? */
+        if (m->pending_hostname) {
+                assert(m->pending_hostname_link);
+                r = link_set_hostname(m->pending_hostname_link, m->pending_hostname);
+                if (r < 0)
+                        log_link_error_errno(m->pending_hostname_link, r, "Failed to set transient hostname to '%s': %m", m->pending_hostname);
+                m->pending_hostname_link = NULL;
+                m->pending_hostname = mfree(m->pending_hostname);
+        }
```
Comment (poettering) [inline src/network/networkd-manager.c]: so, I think link_set_hostname() should be moved out of the "Link" context, and into the "Manager" context. i.e. chnage link_set_hostname() to become manager_set_hostname()... It's a system-wide operation it executes after all, not really a per-link operation...

I'd also do away with the "pending" concept. Simply have manager_set_hostname() do two things: store the hostname parameter in some field Manager::configured_hostname, and try to sync it onto hostnamed. And then, each time we gain a bus connection sync it again. I think this is ultimately simpler and more robust, and effectively the same. After all normally we should only get a bus connection once, and hence it's entirely safe to just do the sync hostname thing each time we connect to the bus...

**Comment by martinpitt** [comment]
Indeed, that makes a lot of sense, will change accordingly. Thanks for the review!

**Comment by martinpitt** [comment]
Reworked. During this I noticed that networkd can set a DHCP-acquired timezone as well, this needs some equivalent fixes (thus adding the "bug" label).

The first commit moves the hostname and timezone setting from Link to Manager, as discussed above; this should have a net-zero effect, but I kept it separate for easier review. The second adds the missing polkit rule for timezone and adds a test for it. The third does the "remember tz/hostname and re-apply after D-Bus connection", and the fourth removes the `After=dbus.service`.

**Comment by martinpitt** [comment]
> See the discussion in #4504 about replacing the 5s retry loop with a signal handler. 

FTR -- I think that would be more elegant, but IMHO it's orthogonal to this PR, so if we still want to do this I'd use a separate PR for this.

File: src/network/networkd-manager.h [hunk 3]
```diff
@@ -63,6 +63,8 @@ struct Manager {
         usec_t network_dirs_ts_usec;
 
         DUID duid;
+        char* dynamic_hostname;
+        char* dynamic_timezone;
 };
```
Comment (poettering) [inline src/network/networkd-manager.h:68]: please free these two fields when destructing the manager object!

**Comment by martinpitt** [comment]
@poettering: Righto, thanks for spotting. Fixed and force-pushed.

**Comment by poettering** [comment]
lgtm! Thanks for working on this!

**Comment by poettering** [comment]
CI failure appears unrelated.

**Pull request merged**